### PR TITLE
feat(language-service): introduce hybrid visitor to locate AST node

### PR DESCRIPTION
**Pull request by kyliau**
This commit introduces two visitors, one for Template AST and the other
for Expression AST to allow us to easily find the node that most closely
corresponds to a given cursor position.

This is crucial because many language service APIs take in a `position`
parameter, and the information returned depends on how well we can find
a good candidate node.

In View Engine implementation of language service, the search for the node
and the processing of information to return the result are strongly coupled.
This makes the code hard to understand and hard to debug because the stack
trace is often littered with layers of visitor calls.

With this new feature, we could test the "searching" part separately and
colocate all the logic (aka hacks) that's required to retrieve an accurate
span for a given node.

Right now, only the most "narrow" node is returned by the main exported
function `findNodeAtPosition`. If needed, we could expose the entire AST
path, or expose other methods to provide more context for a node.

Note that due to limitations in the template AST interface, there are
a few known cases where microsyntax spans are not recorded properly.
This will be dealt with in a follow-up PR.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by josephperrott — APPROVED** [review]
LGTM

Reviewed-For: dev-infra

File: packages/language-service/ivy/test/hybrid_visitor_spec.ts [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+  it('should locate bound attribute key', () => {
+    const {errors, nodes, position} = parse(`<div *ngFor="let item o¦f items"></div>`);
+    expect(errors).toBeUndefined();
+    const node = findNodeAtPosition(nodes, position);
+    expect(isTemplateNode(node!)).toBe(true);
+    expect(node).toBeInstanceOf(t.BoundAttribute);
+    expect((node as t.BoundAttribute).name).toBe('ngForOf');
+  });
+
+  it('should locate bound attribute value', () => {
```
Comment (atscott) [inline packages/language-service/ivy/test/hybrid_visitor_spec.ts:393]: Would it be useful to add a test for the usage of the context variables inside the template?
```
<div *ngFor="let item of items; let i = index">
  {{it¦em}}
  {{i¦}}
</div>
```

**Review by atscott — APPROVED** [review]

File: packages/language-service/ivy/hybrid_visitor.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+class ExpressionVisitor extends e.RecursiveAstVisitor {
+  // Position must be absolute in the source file.
+  constructor(private readonly position: number) {
+    super();
+  }
+
+  visit(node: e.AST, path: Array<t.Node|e.AST>) {
+    if (node instanceof e.ASTWithSource) {
+      node = node.ast;
+    }
```
Comment (ayazhafiz) [inline packages/language-service/ivy/hybrid_visitor.ts:134]: why is this needed?

File: packages/language-service/ivy/hybrid_visitor.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    const {start, end} = node.sourceSpan;
+    // The third condition is to account for the implicit receiver, which should
+    // not be visited.
+    if (start <= this.position && this.position <= end && start < end) {
```
Comment (ayazhafiz) [inline packages/language-service/ivy/hybrid_visitor.ts:138]: what do you think of an explicit check for implicit receiver?

```suggestion
    // The third condition is to account for the implicit receiver, which should
    // not be visited.
    if (start <= this.position && this.position <= end && !(node instanceof e.ImplicitReceiver)) {
```

File: packages/language-service/ivy/hybrid_visitor.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  }
+}
+
+export function isTemplateNode(node: t.Node|e.AST): node is t.Node {
+  // R3 node implements the Node interface so we cannot use instanceof.
+  return node.sourceSpan instanceof ParseSourceSpan;
+}
+
+export function isExpressionNode(node: t.Node|e.AST): node is e.AST {
+  // AST node extends AST so we can use instanceof.
```
Comment (ayazhafiz) [inline packages/language-service/ivy/hybrid_visitor.ts:151]: nit: I think this comment is unncessary

File: packages/language-service/ivy/test/hybrid_visitor_spec.ts [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    expect(node).toBeInstanceOf(t.BoundEvent);
+  });
+
+  it('should locate template bound event value', () => {
+    const {errors, nodes, position} = parse(`<ng-template (click)="f¦oo()"></ng-template>`);
+    expect(errors).toBeUndefined();
+    const node = findNodeAtPosition(nodes, position);
+    expect(node).toBeInstanceOf(e.MethodCall);
+  });
+
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/hybrid_visitor_spec.ts:171]: Could you add a test for banana-in-a-box attributes? I guess for now they would only return a bound event or bound attribute, which seems okay.

File: packages/language-service/ivy/test/hybrid_visitor_spec.ts [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    expect(node).toBeInstanceOf(t.BoundAttribute);
+  });
+
+  it('should locate bound attribute value', () => {
+    const {errors, nodes, position} = parse(`<test-cmp [foo]="b¦ar"></test-cmp>`);
+    expect(errors).toBeUndefined();
+    const node = findNodeAtPosition(nodes, position);
+    expect(isExpressionNode(node!)).toBe(true);
+    expect(node).toBeInstanceOf(e.PropertyRead);
+  });
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/hybrid_visitor_spec.ts:98]: nit: should this go in the expression AST spec? I actually like it better here, but curious on your thoughts.

File: packages/language-service/ivy/test/hybrid_visitor_spec.ts [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+    expect(node).toBeInstanceOf(e.LiteralMap);
+  });
+
+  it('should locate conditional', () => {
+    const {errors, nodes, position} = parse(`{{ cond ?¦ true : false }}`);
+    expect(errors).toBeUndefined();
+    const node = findNodeAtPosition(nodes, position);
+    expect(isExpressionNode(node!)).toBe(true);
+    expect(node).toBeInstanceOf(e.Conditional);
+  });
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/hybrid_visitor_spec.ts:337]: Could you add tests for property access, safe property access, and binary expression?

File: packages/language-service/ivy/test/hybrid_visitor_spec.ts (hunk 1 above)
Comment (kyliau) [reply packages/language-service/ivy/test/hybrid_visitor_spec.ts:480]: I added a test for context variables inside a template.
In addition, your example actually reminded me to test one more scenario:
```
<div *ngFor="let item of items; let i = index">
                                ^^^^^^^^^^^^^
</div>
```
Currently there is no way to distinguish LHS and RHS of the variable declaration.
For LHS we don't care, but we have to be able to detect RHS because it points to `NgForOfContext.index`.
I left a TODO for this case.

**Review by ayazhafiz — APPROVED** [review]
+me after comment resolved. Awesome change!

**Review by ivanwonder — COMMENTED** [review]
amazing hybrid visitor
